locallib.php

```diff
@@ -475,5 +477,49 @@ function auth_saml2_get_default_idp() {
     return $defaultidp;
 }
 
+/**
+ * This helper function processes the regenerate form.
+ * Moved here so we can use it in a unit test.
+ *
+ * @param $fromform
+ * @return void|string
+ */
+function auth_saml2_process_regenerate_form($fromform) {
+    global $CFG, $USER;
+    $dn = array(
+        'commonName' => substr($fromform->commonname, 0, 64),
+        'countryName' => $fromform->countryname,
+        'emailAddress' => $fromform->email,
+        'localityName' => $fromform->localityname,
+        'organizationName' => $fromform->organizationname,
+        'stateOrProvinceName' => $fromform->stateorprovincename,
+        'organizationalUnitName' => $fromform->organizationalunitname,
+    );
+    $numberofdays = $fromform->expirydays;
+
+    $saml2auth = new \auth_plugin_saml2();
+    $error = create_certificates($saml2auth, $dn, $numberofdays);
+
+    if (!$error) {
+        // Successfully regenerated cert so emit the cert_regenerated event.
+        $eventdata = [
+            'reason' => "regenerated in saml settings page",
+            'userid' => $USER->id,
+        ];
+        cert_regenerated::create(['other' => $eventdata])->trigger();
+    }
+
+    // Also refresh the SP metadata as well.
+    $file = $saml2auth->get_file_sp_metadata_file();
+    @unlink($file);
+
+    // Can't redirect in a unit test.
+    if (empty($error) && !PHPUNIT_TEST) {
```

Comment:
If code is aware of unit tests then usually thats a code smell of the interface not being right.

A better solution would be to redirect in the calling code, if and only if the result was null. And then you can cleanly unit tests without this hack

Reply:
@brendanheywood  I have moved that redirect now